### PR TITLE
Transactional migration (#70)

### PR DESCRIPTION
**Pull request by phips28**
Issue: https://github.com/timgit/pg-boss/issues/70

Migration: use pg transaction to avoid inconsistency. In case one query fails, or the node process restarts during the query is running, it will automatically rollback. 

Param `poolSize`: also use `max` as valid config parameter (its the underlying pg-pool param) 

Tests: changed timeouts to work with my remote database; unfortunately the `speedTest.js` doesnt work with my remote database 👎  but works on travis 👍

### CODE REVIEW

**Comment by timgit** [comment]
@phips28 Thanks for this!  🙇 

btw, I downgraded bluebird to a dev dependency, which actually introduces a use case that Travis didn't catch since it would only show up with the --production arg or NODE_ENV var.

Also, all the commands don't need an extra semicolon, do they?  `join(';')` should take care of that, right?

**Comment by phips28** [comment]
> Also, all the commands don't need an extra semicolon, do they? join(';') should take care of that, right?

Yes thats true, I first added the `;` in every query, but then I realized that the migration.js would need the `;` too at the end adn I was too lazy to edit all. So I used the join(';').

I removed it

File: src/contractor.js [hunk 1]
```diff
@@ -2,6 +2,7 @@ const assert = require('assert');
 const plans = require('./plans');
 const migrations = require('./migrations');
 const schemaVersion = require('../version.json').schema;
+const Promise = require('bluebird');
```
Comment (timgit) [inline src/contractor.js]: Remove this

File: src/contractor.js [hunk 2]
```diff
@@ -46,10 +47,13 @@ class Contractor {
...
 
   create(){
-    let promises = plans.create(this.config.schema).map(command => () => this.db.executeSql(command));
-
-    return this.promiseEach(promises)
-      .then(() => this.db.executeSql(plans.insertVersion(this.config.schema), [schemaVersion]));
+    // use transaction, in case one query fails, it will automatically rollback to avoid inconsistency
+    let queryInTransaction = `
+    BEGIN;
+    ${plans.create(this.config.schema).join(';')}
```
Comment (timgit) [inline src/contractor.js]: Add semicolon after the join() to terminate last command

File: src/contractor.js [hunk 3]
```diff
@@ -46,10 +47,13 @@ class Contractor {
...
+    ${plans.insertVersion(this.config.schema).replace('$1', `'${schemaVersion}'`)}
```
Comment (timgit) [inline src/contractor.js]: Add semicolon after this as well to terminate command

File: src/contractor.js [hunk 4]
```diff
@@ -82,23 +86,21 @@ class Contractor {
...
       return Promise.reject(new Error(errorMessage));
     }
-
-    let promises = migration.commands.map(command => () => this.db.executeSql(command));
-
-    return this.promiseEach(promises)
+    // use transaction, in case one query fails, it will automatically rollback to avoid inconsistency
+    let queryInTransaction = `
+    BEGIN;
+    ${migration.commands.join(';')}
```
Comment (timgit) [inline src/contractor.js]: Add semicolon after the join() to terminate last command

File: test/databaseTest.js [hunk 5]
```diff
@@ -108,7 +104,47 @@ describe('database', function(){
...
     function countConnections(db) {
-      return db.executeSql('SELECT count(*) as connections FROM pg_stat_activity')
+      return db.executeSql('SELECT count(*) as connections FROM pg_stat_activity WHERE application_name=$1',
+        [boss.db.config.application_name])
+        .then(result => parseFloat(result.rows[0].connections));
+    }
+
+  });
+
+  it('connection count does not exceed configured pool size with `max`', function(finished){
```
Comment (timgit) [inline test/databaseTest.js]: I'd prefer to move this into a config test (does poolSize == max) instead of duplicating the entire connection pooling test

Comment (phips28) [reply test/databaseTest.js]: added new tests, do you like it that way? should we remove the pooling test?

Comment (timgit) [reply test/databaseTest.js]: Thanks!  I'll go ahead and merge and take a stab at the migration failure rolllback test

**Pull request merged**